src/libsync/syncfilestatustracker.cpp

```diff
@@ -0,0 +1,182 @@
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#include "syncfilestatustracker.h"
+#include "syncengine.h"
+#include "syncjournaldb.h"
+#include "syncjournalfilerecord.h"
+
+namespace OCC {
+
+static SyncFileStatus::SyncFileStatusTag lookupProblem(const QString &pathToMatch, const std::set<Problem> &set)
+{
+    for (auto it = set.cbegin(); it != set.cend(); ++it) {
+        qDebug() << Q_FUNC_INFO << pathToMatch << it->severity << it->path;
+        auto problemPath = it->path;
+        if (problemPath == pathToMatch)
+            return it->severity;
+        else if (it->severity == SyncFileStatus::StatusError && problemPath.startsWith(pathToMatch))
+            return SyncFileStatus::StatusWarning;
+    }
+    return SyncFileStatus::StatusNone;
+}
+
+/**
+ * Whether this item should get an ERROR icon through the Socket API.
+ *
+ * The Socket API should only present serious, permanent errors to the user.
+ * In particular SoftErrors should just retain their 'needs to be synced'
+ * icon as the problem is most likely going to resolve itself quickly and
+ * automatically.
+ */
+static inline bool showErrorInSocketApi(const SyncFileItem& item)
+{
+    const auto status = item._status;
+    return status == SyncFileItem::NormalError
+        || status == SyncFileItem::FatalError
+        || item._hasBlacklistEntry;
+}
+
+static inline bool showWarningInSocketApi(const SyncFileItem& item)
+{
+    const auto status = item._status;
+    return status == SyncFileItem::FileIgnored
+        || status == SyncFileItem::Conflict
+        || status == SyncFileItem::Restoration;
+}
+
+SyncFileStatusTracker::SyncFileStatusTracker(SyncEngine *syncEngine)
+    : _syncEngine(syncEngine)
+{
+    connect(syncEngine, SIGNAL(aboutToPropagate(SyncFileItemVector&)),
+              this, SLOT(slotAboutToPropagate(SyncFileItemVector&)));
+    connect(syncEngine, SIGNAL(itemCompleted(const SyncFileItem&, const PropagatorJob&)),
+            this, SLOT(slotItemCompleted(const SyncFileItem&)));
+}
+
+
+SyncFileStatus SyncFileStatusTracker::fileStatus(const QString& systemFileName)
+{
+    QString fileName = systemFileName.normalized(QString::NormalizationForm_C);
+    if( fileName.endsWith(QLatin1Char('/')) ) {
+        fileName.truncate(fileName.length()-1);
+        qDebug() << "Removed trailing slash: " << fileName;
+    }
+
+    SyncFileItem* item = _syncEngine->findSyncItem(fileName);
+    if (item)
+        return fileStatus(*item);
+
+    // If we're not currently syncing that file, look it up in the database to know if it's shared
+    SyncJournalFileRecord rec = _syncEngine->journal()->getFileRecord(fileName);
+    if (rec.isValid())
+        return fileStatus(rec.toSyncFileItem());
+
+    // Must be a new file, wait for the filesystem watcher to trigger a sync
```

Comment:
I tried it and it felt right that no icon is shown until the sync kicks in (which took 1-2 seconds). It make the behavior consistent with the rest that if the icon is blue, it's because the sync is happening, and this would always be the case.

Otherwise, we would show a blue icon for new files, but the tray icon would stay green, and we will keep showing the OK status for modified files unless we implement the tainted folder logic.

Personally I feel that if the tray icon shown as syncing, the files should also not be shown as "syncing".